tsdb/index/tsi1/series_file.go

```diff
@@ -0,0 +1,446 @@
+
+// ErrSeriesOverflow is returned when too many series are added to a series writer.
+var ErrSeriesOverflow = errors.New("series overflow")
+
+// Series list field size constants.
+const SeriesIDSize = 4
+
+// Series flag constants.
+const (
+	// Marks the series as having been deleted.
+	SeriesTombstoneFlag = 0x01
+
+	// Marks the following bytes as a hash index.
+	// These bytes should be skipped by an iterator.
+	SeriesHashIndexFlag = 0x02
+)
+
+const DefaultMaxSeriesFileSize = 32 * (1 << 30) // 32GB
+
+// MaxSeriesFileHashSize is the maximum number of series in a single hash.
+const MaxSeriesFileHashSize = (1048576 * LoadFactor) / 100
+
+// SeriesFile represents the section of the index that holds series data.
+type SeriesFile struct {
+	path string
+	data []byte
+	file *os.File
+	w    *bufio.Writer
+
+	offset  uint32
+	hashMap *rhh.HashMap
+
+	// MaxSize is the maximum size of the file.
+	MaxSize int64
+}
+
+// NewSeriesFile returns a new instance of SeriesFile.
+func NewSeriesFile(path string) *SeriesFile {
+	return &SeriesFile{
+		path:    path,
+		MaxSize: DefaultMaxSeriesFileSize,
+	}
+}
+
+// Open memory maps the data file at the file's path.
+func (f *SeriesFile) Open() error {
+	// Open file handler for appending.
+	file, err := os.OpenFile(f.path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		return err
+	}
+	f.file = file
+
+	// Ensure header byte exists.
+	f.offset = 0
+	if fi, err := file.Stat(); err != nil {
+		return err
+	} else if fi.Size() > 0 {
+		f.offset = uint32(fi.Size())
+	} else {
+		if _, err := f.file.Write([]byte{0}); err != nil {
+			return err
+		}
+		f.offset = 1
+	}
+
+	// Wrap file write a bufferred writer.
+	f.w = bufio.NewWriter(f.file)
+
+	// Memory map file data.
+	data, err := mmap.Map(f.path, f.MaxSize)
+	if err != nil {
+		return err
+	}
+	f.data = data
+
+	// Index all series.
+	if err := f.reindex(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Close unmaps the data file.
+func (f *SeriesFile) Close() error {
+	if f.data != nil {
+		if err := mmap.Unmap(f.data); err != nil {
+			return err
+		}
+		f.data = nil
+	}
+
+	if f.file != nil {
+		if err := f.file.Close(); err != nil {
+			return err
+		}
+		f.file = nil
+	}
+
+	f.hashMap = nil
+
+	return nil
+}
+
+// Path returns the path to the file.
+func (f *SeriesFile) Path() string { return f.path }
+
+// reindex iterates over all series and builds a hash map index.
+func (f *SeriesFile) reindex() error {
+	m := rhh.NewHashMap(rhh.DefaultOptions)
+
+	// Series data begins with an offset of 1.
+	data := f.data[1:f.offset]
+	offset := uint32(1)
+
+	for len(data) > 0 {
+		var key []byte
+		key, data = ReadSeriesKey(data)
+
+		m.Put(key, offset)
+		offset += uint32(len(key))
+	}
+
+	f.hashMap = m
+
+	return nil
+}
+
+// CreateSeriesIfNotExists creates series if it doesn't exist. Returns the offset of the series.
+func (f *SeriesFile) CreateSeriesIfNotExists(name []byte, tags models.Tags, buf []byte) (offset uint32, err error) {
```

Comment:
Is the offset returned as the series ID stable?  Will it change each time the hash is re-indexed?  If so, then I won't be able to use it in TSM files as I need a stable ID.

Reply:
Also, I need the field type and field name to be part of the key as well.

Reply:
The offset is from the series file which is append-only so it will be stable. I'll add field name/type as well.

Reply:
Cool